cep-20.2.md

```diff
@@ -0,0 +1,831 @@
+and types in a single document and document them. This is part of a larger effort (see [CEP-13](/cep-13.md) for the new YAML syntax).
+
+## History
+
+A discussion was started on what a new recipe spec could or should look like.
+The fragments of this discussion can be found here:
+https://github.com/mamba-org/conda-specs/blob/master/proposed_specs/recipe.md
+The reason for a new spec are:
+
+- Make it easier to parse ("pure yaml"). conda-build uses a mix of comments and
+  jinja to achieve a great deal of flexibility, but it's hard to parse the
+  recipe with a computer
+- iron out some inconsistencies around multiple outputs (build vs. build/script
+  and more)
+- remove any need for recursive parsing & solving
+- cater to needs for automation and dependency tree analysis via a determinstic
+  format
+
+## Major differences with conda-build
+
+Outputs and top-level package are mutually exclusive, and outputs have exactly the same structure as top-level keys.
+If outputs exist, the top-level keys are 'merged' with the output keys (e.g. for the about section).
+
+# Format
+
+## Schema version
+
+The implicit version of the YAML schema for a recipe is an integer 1.
+To discern between the "old" format and the new format, we utilize the file name.
+The old format is `meta.yaml` and the new format is `recipe.yaml`.
+The version can be explicitly set by adding a `schema_version` key to the recipe.
+
+```yaml
+# optional, since implicitly defaults to 1
+schema_version: 1 # integer
+```
+
+To benefit from autocompletion, and other LSP features in editors, we can add a schema URL to the recipe.
+
+```yaml
+# yaml-language-server: $schema=https://raw.githubusercontent.com/prefix-dev/recipe-format/main/schema.json
+```
+
+## Context section
+
+```yaml
+# The context dictionary defines arbitrary key-value pairs for Jinja interpolation
+# and replaces the {% set ... %} commands commonly used in recipes
+context:
+  variable: test
+  # note that we can reference previous values, that means that they are rendered in order
+  other_variable: test_${{ variable }}
+```
+
+## Package section
+
+```yaml
+package:
+  # The name of the package
+  name: string
+  # The version of the package, following the conda version spec
+  # Note that versions are always strings (1.23 is _not_ a valid version since it is a `float`. Needs to be quoted.)
+  version: string
+```
+
+## Build section
+
+```yaml
+build:
+  # the build number
+  number: Option<integer>, defaults to 0
+
+  # the build string. This is usually ommited (can use `${{ hash }}` variable here)
+  string: Option<string>, defaults to a build string made from "package hash & build number"
+
+  # A list of conditions under which to skip the build of this package (they are joined by `or`)
+  skip: [list of expressions]
+
+  # wether the package is a noarch package, and if yes, wether it is "generic" or "python"
+  noarch: Option<"generic" | "python">
```

Comment:
It would be a good opportunity to start to split `noarch: python` into `noarch` and `package-type: python` (not that we have to use the `package-type` naming, but you get the gist).

This separation of concerns should be done upstream in `conda` etc. -- but that needs changing of index data format.
It would be good to prepare recipes for such separation, though.
For now, an package building tool would still output `noarch: python` -- but it could simply switch to a new index format once that is available (so that we don't have to change recipes again at that point).

Reply:
I think this needs more careful thoughts. Lets do this in a future CEP. 

We also would like to support Python3 stable ABI packages - they are not `noarch` but still need special hadnling to move the contents to the correct `site-packages` folder.

Reply:
PS this proposal already got rid of the `noarch: true` option :)

Reply:
> I think this needs more careful thoughts. Lets do this in a future CEP.

I would be fine with postponing this -- but it would also be good if we can handle this before the new recipe format sees wider use.

> We also would like to support Python3 stable ABI packages - they are not `noarch` but still need special hadnling to move the contents to the correct `site-packages` folder.

Yes, that makes sense.
As such it would be good to move this into a package-type-specific section (you already proposed `python:` -- which is a good thing!).

Reply:
I think even later we can easily mechanically update the recipes to what you are envisioning.